src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+
+
+                // if there are more than one return in the method, add yield break;
+                if (memberDeclarationNode.DescendantNodes().OfType<ReturnStatementSyntax>().Count() != 1)
+                {
+                    var yieldBreakStatement = SyntaxFactory.YieldStatement(SyntaxKind.YieldBreakStatement);
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, statements.Concat(new[] { yieldBreakStatement }));
+                }
+                else
+                {
+                    documentUpdateInfo = new DocumentUpdateInfo(returnStatement, statements);
+                }
+
+                return true;
+            }
+
+            private static SyntaxNode FindParentMemberDeclarationNode(SyntaxNode node)
+            {
+                while (node != null)
+                {
+                    switch (node.Kind())
+                    {
+                        case SyntaxKind.MethodDeclaration:
+                        case SyntaxKind.LocalFunctionStatement:
+                        case SyntaxKind.PropertyDeclaration:
+                        case SyntaxKind.AnonymousMethodExpression:
+                            return node;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return null;
+            }
+
+            private bool TryCreateStackFromQueryExpression(out QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                queryExpressionProcessingInfo = new QueryExpressionProcessingInfo(_source.FromClause);
+                return TryProcessQueryBody(_source.Body, queryExpressionProcessingInfo);
+            }
+
+            private StatementSyntax[] GenerateStatements(
+                Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                StatementSyntax statement = default;
+                var stack = queryExpressionProcessingInfo.Stack;
+                // Executes syntax building methods from bottom to the top of the tree.
+                // Process last clause
+                if (stack.Any())
+                {
+                    var node = stack.Pop();
+                    if (node is SelectClauseSyntax selectClause)
+                    {
+                        statement = WrapWithBlock(leafExpressionCreationMethod(selectClause.Expression));
+                    }
+                    else
+                    {
+                        throw new ArgumentException("Last node must me the select clause");
+                    }
+                }
+
+                // Process all other clauses
+                List<StatementSyntax> statements = new List<StatementSyntax>();
+                while (stack.Any())
+                {
+                    statement = ProcessClause(stack.Pop(), statement, out StatementSyntax extraStatement);
```

Comment:
>statement = ProcessClause(stack.Pop(), statement, out StatementSyntax extraStatement); [](start = 20, length = 86)

It looks like, if there are extra statements, collection from the top most from should go into extra statement too.